File: frontend/lib/types/breakdown.ts
// ─── Explanation styles ───────────────────────────────────────────────────────

export type ExplanationStyle =
  | "simple"
  | "analogy"
  | "example"
  | "prerequisite"
  | "diagram"
  | "eli5";

export const EXPLANATION_STYLES: {
  id: ExplanationStyle;
  label: string;
  description: string;
  icon: string;
}[] = [
  { id: "simple",       label: "Simpler",     description: "Plain language, no jargon",          icon: "Feather"   },
  { id: "analogy",      label: "Analogy",      description: "Relatable comparison",               icon: "GitCompare" },
  { id: "example",      label: "Example",      description: "Concrete real-world scenario",       icon: "Lightbulb" },
  { id: "prerequisite", label: "Prerequisites",description: "What to know first",                 icon: "BookOpen"  },
  { id: "diagram",      label: "Diagram",      description: "Visual / structural description",    icon: "LayoutTemplate" },
  { id: "eli5",         label: "ELI5",         description: "Like I'm 5 years old",               icon: "Baby"      },
];

// ─── Languages ────────────────────────────────────────────────────────────────

export interface SupportedLanguages {
  languages: Record<string, string>; // { code: name }
}

export const DEFAULT_LANGUAGE = "en";

// ─── Explanation ──────────────────────────────────────────────────────────────

export interface Explanation {
  id: string;
  concept_id: string;
  style: ExplanationStyle;
  language: string;
  content: string;
  source_quote: string | null;
  timestamp_start: number | null;
  audio_url: string | null;
  audio_duration_ms: number | null;
  cached: boolean;
  generated_at: string | null;
}

// ─── Audio ────────────────────────────────────────────────────────────────────

export interface AudioResult {
  explanation_id: string;
  audio_url: string;
  duration_ms: number | null;
  cached: boolean;
  error: string | null;
}

// ─── Revisit recommendations ──────────────────────────────────────────────────

export interface RevisitRecommendation {
  concept_id: string;
  concept_name: string;
  reason: string;
  priority: number;
  timestamp_start: number | null;
  exam_likelihood: number;
  importance: string;
}

// ─── Request payloads ─────────────────────────────────────────────────────────

export interface ExplainRequest {
  concept_id: string;
  style: ExplanationStyle;
  language: string;
}

export interface AudioRequest {
  explanation_id: string;
}

File: backend/app/schemas/breakdown.py
from datetime import datetime
from pydantic import BaseModel, Field
from app.breakdown.schemas import ExplanationStyle, SUPPORTED_LANGUAGES


# ── Request schemas ───────────────────────────────────────────────────────────

class ExplainRequest(BaseModel):
    concept_id: str
    style: ExplanationStyle = ExplanationStyle.SIMPLE
    language: str = Field(default="en", pattern=r"^[a-z]{2}$")

    def validated_language(self) -> str:
        return self.language if self.language in SUPPORTED_LANGUAGES else "en"


class AudioRequest(BaseModel):
    explanation_id: str


# ── Response schemas ──────────────────────────────────────────────────────────

class ExplanationRead(BaseModel):
    id: str
    concept_id: str
    style: str
    language: str
    content: str
    source_quote: str | None = None
    timestamp_start: float | None = None
    audio_url: str | None = None         # set when audio has been generated
    audio_duration_ms: int | None = None
    cached: bool = False
    generated_at: datetime | None = None

    model_config = {"from_attributes": True}


class AudioRead(BaseModel):
    explanation_id: str
    audio_url: str
    duration_ms: int | None = None
    cached: bool = False
    error: str | None = None


class RevisitRecommendationRead(BaseModel):
    concept_id: str
    concept_name: str
    reason: str
    priority: int
    timestamp_start: float | None = None
    exam_likelihood: float = 0.5
    importance: str = "supporting"


class SupportedLanguagesRead(BaseModel):
    languages: dict[str, str]   # {code: name}
